src/physics/boundaryConditions/PeriodicBoundary.cpp

```diff
@@ -0,0 +1,230 @@
+{
+    const LennardJonesDomainSimulation& LGDSim =
+        static_cast<const LennardJonesDomainSimulation&>(simulation);
+
+    // reset the insertion index, to reuse already created particles
+    insertionIndex = 0;
+    for (CellIndex boundaryCellIndex : LGDSim.getGrid().boundaryCellIterator(position)) {
+        std::vector<Position> boundarySides = coordinateToPosition(
+            boundaryCellIndex, LGDSim.getGrid().getGridDimensions(), !is2D, false);
+
+        // Only if translations need to be applied i.e. there exists an entry
+        if (translationMap.find(boundarySides) == translationMap.end())
+            continue;
+
+        for (auto& particle :
+             LGDSim.getGrid()
+                 .cells[boundaryCellIndex[0]][boundaryCellIndex[1]][boundaryCellIndex[2]]
+                 ->getParticles()) {
+            for (PeriodicBoundShifts& shifts : translationMap.at(boundarySides)) {
+                std::array<double, 3> pos = particle.get().getX();
+                std::array<double, 3> haloPosition = pos + shifts.first;
+
+                // Ugly but we cannot simply add as size_t != int
+                CellIndex haloCellIndex { (size_t)((int)boundaryCellIndex[0] + shifts.second[0]),
+                                          (size_t)((int)boundaryCellIndex[1] + shifts.second[1]),
+                                          (size_t)((int)boundaryCellIndex[2] + shifts.second[2]) };
+
+                if (LGDSim.getGrid().determineCellType(haloCellIndex) != CellType::Halo) {
+                    spdlog::error("Halo cell index is not a halo cell");
+                    exit(EXIT_FAILURE);
+                }
+
+                if (insertionIndex < insertedParticles.size()) {
+                    // simply update, if there is already a particle to reuse
+                    insertedParticles.at(insertionIndex)->setX(haloPosition);
+                    insertedParticles.at(insertionIndex)->setV({ 0, 0, 0 });
+                    insertedParticles.at(insertionIndex)->setF({ 0, 0, 0 });
+                    insertedParticles.at(insertionIndex)->setOldF({ 0, 0, 0 });
+                    insertedParticles.at(insertionIndex)->setM(particle.get().getM());
+                    insertedParticles.at(insertionIndex)->setType(particle.get().getType());
+                    insertedParticles.at(insertionIndex)->setActivity(false);
+                } else {
+                    Particle haloParticle(
+                        haloPosition, { 0, 0, 0 }, particle.get().getM(), particle.get().getType());
+                    haloParticle.setActivity(false);
+                    insertedParticles.push_back(std::make_unique<Particle>(haloParticle));
+                }
+
+                // add the particle to the halo cell
+                LGDSim.getGrid()
+                    .cells[haloCellIndex[0]][haloCellIndex[1]][haloCellIndex[2]]
+                    ->addParticle(*insertedParticles.at(insertionIndex++).get());
```

Comment:
```suggestion
                    ->addParticle(*insertedParticles.at(insertionIndex++));
```
Dereferencing also works on smart pointers directly